crates/oxc_linter/src/rules/typescript/no_this_alias.rs

```diff
@@ -0,0 +1,213 @@
+#[error("typescript-eslint(no-this alias): Unexpected aliasing of 'this' to local variable.")]
+#[diagnostic(severity(error), help("Unexpected aliasing of 'this' to local variable."))]
+struct NoThisAliasDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Error, Diagnostic)]
+#[error(
+    "typescript-eslint(no-this alias): Unexpected aliasing of members of 'this' to local variables."
+)]
+#[diagnostic(severity(error), help("Unexpected aliasing of members of 'this' to local variables."))]
+struct NoThisDestructureDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Clone)]
+pub struct NoThisAlias {
+    allow_destructuring: bool,
+    allow_names: Vec<String>,
+}
+
+impl Default for NoThisAlias {
+    fn default() -> Self {
+        Self { allow_destructuring: true, allow_names: vec![] }
+    }
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallow unnecessary constraints on generic types.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Generic type parameters (<T>) in TypeScript may be "constrained" with an extends keyword.
+    /// When no extends is provided, type parameters default a constraint to unknown. It is therefore redundant to extend from any or unknown.
+    ///
+    /// the rule doesn't allow const {allowedName} = this
+    /// this is to keep 1:1 with eslint implementation
+    /// sampe with obj.<allowedName> = this
+    /// ```
+    NoThisAlias,
+    correctness
+);
+
+impl Rule for NoThisAlias {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let obj = value.get(0);
+        let allowed_names = value
+            .get(0)
+            .and_then(|v| v.get("allow_names"))
+            .and_then(serde_json::Value::as_array)
+            .unwrap_or(&vec![])
+            .iter()
+            .map(serde_json::Value::as_str)
+            .filter(std::option::Option::is_some)
+            .map(|x| x.unwrap().to_string())
+            .collect::<Vec<String>>();
+
+        Self {
+            allow_destructuring: obj
+                .and_then(|v| v.get("allow_destructuring"))
+                .and_then(serde_json::Value::as_bool)
+                .unwrap_or_default(),
+            allow_names: allowed_names,
+        }
+    }
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        match node.kind() {
+            AstKind::VariableDeclarator(decl) => {
+                if decl.init.is_none() {
+                    return;
+                }
+
+                if let Some(init) = &decl.init && !rhs_is_this_reference(init) {
+                    return;
+                }
+
+                if self.allow_destructuring
+                    && !matches!(decl.id.kind, BindingPatternKind::BindingIdentifier(_))
+                {
+                    return;
+                }
+
+                if let BindingPatternKind::BindingIdentifier(identifier) = &decl.id.kind {
+                    if !self.allow_names.contains(&identifier.name.to_string()) {
```

Comment:
We don't need the `to_string` here, it's an extra allocation we can avoid.

Reply:
Will do!